### PR TITLE
Orbit iteration fix when no loaded data (non-mlt orbits)

### PR DESCRIPTION
**Pull request by rstoneback**

### CODE REVIEW

**Comment by rstoneback** [comment]
This branch was made before the SuperMag fix was in, hence the tests are failing.

**Comment by rstoneback** [comment]
Merged in latest develop to help out with tests.

File: pysat/instruments/omni_hro.py [hunk 1]
```diff
@@ -222,8 +222,8 @@ def calculate_imf_steadiness(inst, steady_window=15, min_window_frac=0.75,
         inst['clock_angle'].rolling(min_periods=min_wnum,
                                     window=steady_window,
                                     center=True).apply(pystats.nan_circstd,
-                                                       kwargs=circ_kwargs,
-                                                       raw=True)
+                                                       kwargs=circ_kwargs)
+                                                       #raw=True)
```
Comment (jklenzing) [inline pysat/instruments/omni_hro.py]: Added `raw=True` back in #173.  Now required as part of pandas.  The True option sends the ndarray directly to the `rolling.apply` function.

From https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.window.Rolling.apply.html
```
The raw parameter is required and will show a FutureWarning if not passed. In the future raw will default to False.

New in version 0.23.0.
```

File: pysat/_orbits.py [hunk 2]
```diff
@@ -344,6 +344,8 @@ def _polarBreaks(self):
             # print 'Time Gap'
 
         # create orbitbreak index, ensure first element is always 0
+        if len(ind) == 0:
+            ind = np.array([0])
```
Comment (jklenzing) [inline pysat/_orbits.py]: Can you add a unit test for the 0 length case?

File: pysat/tests/test_orbits.py [hunk 3]
```diff
@@ -1,11 +1,26 @@
...
 import pysat
 
 
+class TestOrbitsUserInterface():
+
+    @raises(NameError)
+    def test_orbit_wo_info(self):
+        self.testInst = pysat.Instrument('pysat', 'testing',
+                                         clean_level='clean',
+                                         orbit_info=info, update_files=True)
```
Comment (jklenzing) [inline pysat/tests/test_orbits.py]: Looking over coveralls, this doesn't add anything to the coverage.  I think the name error is generated deeper in one of the packages, not in our code.  Should we keep it?

Comment (jklenzing) [reply pysat/tests/test_orbits.py]: Went ahead and pulled this extra test out.

Comment (jklenzing) [reply pysat/tests/test_orbits.py]: Went ahead and pulled this extra test out.

File: pysat/instruments/omni_hro.py (hunk 1 above)
Comment (jklenzing) [reply pysat/instruments/omni_hro.py]: fixed

File: pysat/_orbits.py [hunk 4]
```diff
@@ -381,6 +383,8 @@ def _orbitNumberBreaks(self):
             orbit_index.append(idx[0])
 
         # create orbitbreak index, ensure first element is always 0
+        if len(orbit_index) == 0:
+            orbit_index[0] = 0
```
[the remark this answers is not in the document]
Comment (jklenzing) [reply pysat/_orbits.py]: fixed

**Comment by jklenzing** [comment]
Updated the minor bits, still needs a unit test to catch the bug this fixed in case it reappears.

**Comment by rstoneback** [comment]
Looking at the code I need to generate a situation where there is some data, but not a whole orbits worth. For polar orbit no equatorial crossings.

**Comment by rstoneback** [comment]
Tests are failing on TIMED_SABER downloads.

**Comment by jklenzing** [comment]
> Tests are failing on TIMED_SABER downloads.

Not sure why. Waiting to see the results of the recent develop merge to see if this hits there as well. May be a change on cdaweb’s side.

**Comment by jklenzing** [comment]
Failing on develop too. Will look into potential cdaweb updates tonight.

**Comment by jklenzing** [comment]
Since TIMED_SABER error was part of a change to the external ecosystem, opened #239 as a bug fix.

**Comment by jklenzing** [comment]
Running some local tests.  Looks like the `test_load_orbits_w_empty_data` I added should be removed. If you try to run `.next()` on an empty dataset, it does and should error.  Not sure if we need to check for this.

**Comment by jklenzing** [comment]
Added some pep8!

**Comment by rstoneback** [comment]
I think that test should remain. Makes sure the feature continues to work in the future,

**Comment by jklenzing** [comment]
If we do, the unit tests need to be changed to catch `StopIteration`.  Pulled those lines out back in 
f556e8d.  They were causing about 20 errors since there was no data to split orbits over.  Example:
```
======================================================================
ERROR: pysat.tests.test_orbits.TestOrbitsGappyOrbitNumData.test_load_orbits_w_empty_data
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jklenzing/anaconda3/lib/python3.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jklenzing/code/pysat/pysat/tests/test_orbits.py", line 158, in test_load_orbits_w_empty_data
    self.testInst.orbits.next()
  File "/Users/jklenzing/code/pysat/pysat/_orbits.py", line 684, in next
    # possible
  File "/Users/jklenzing/code/pysat/pysat/_instrument.py", line 1710, in next
    raise StopIteration(''.join(('File list is empty. ',
StopIteration: File list is empty. Nothing to be done.
-------------------- >> begin captured stdout << ---------------------
No pysat testing data for 01 January 1959

--------------------- >> end captured stdout << ----------------------
```

**Comment by jklenzing** [comment]
Added empty case back in.

Still errors on `test_less_than_one_orbit_of_data_four_ways_two_days` and `test_less_than_one_orbit_of_data_two_ways`.

**Comment by jklenzing** [comment]
Suggest we turn off (comment out) the two additional unit test cases for the time being.

File: pysat/tests/test_orbits.py [hunk 5]
```diff
@@ -128,7 +138,80 @@ def teardown(self):
...
+        self.testInst.custom.add(filter_data, 'modify')
+        self.testInst.load(2009, 1)
+        # starting from no orbit calls next loads first orbit
+        self.testInst.orbits.next()
+        # store comparison data
+        saved_data = self.testInst.data.copy()
+        self.testInst.load(2009, 1)
+        self.testInst.orbits[0]
+        print('0 ', saved_data.index, self.testInst.index)
+        assert all(self.testInst.data == saved_data)
```
Comment (jklenzing) [inline pysat/tests/test_orbits.py]: In the current test setup, `saved_data` contains data from 01-01-2009, `self.testInst.data` contains data from 12-31-2008.  Not sure what is up here.

File: pysat/tests/test_orbits.py [hunk 6]
```diff
@@ -128,7 +138,80 @@ def teardown(self):
...
+        self.testInst.orbits[0]
+        if self.testInst.orbits.num == 1:
+            # equivalence only when only one orbit
+            # some test settings can violate this assumption
+            print('1 ', saved_data.index, self.testInst.index)
+            assert all(self.testInst.data == saved_data)
+        self.testInst.load(2009, 4)
+        self.testInst.orbits[0]
+        print('2 ', saved_data.index, self.testInst.index)
+        assert all(self.testInst.data == saved_data)
```
Comment (jklenzing) [inline pysat/tests/test_orbits.py]: Similar error at this point

File: pysat/tests/test_orbits.py (hunk 5 above)
Comment (rstoneback) [reply pysat/tests/test_orbits.py]: The orbits iterator goes past day/file boundaries to load complete orbits.

File: pysat/tests/test_orbits.py (hunk 6 above)
Comment (rstoneback) [reply pysat/tests/test_orbits.py]: New push passes tests locally. I also improved the date check.

File: pysat/_orbits.py (hunk 2 above)
Comment (jklenzing) [reply pysat/_orbits.py]: Lines covered in latest build.  Thanks!

https://coveralls.io/builds/24314833/source?filename=pysat%2F_orbits.py

**Comment by rstoneback** [comment]
Added more tests to cover basic errors. Turns out my expect statements in the various orbit index determination methods were looking for the wrong error. Should've been KeyError instead of ValueError.

**Comment by rstoneback** [comment]
Thanks Jeff and to the #TheEyeOfBurrell (@abubrrell) for spotting gaps in testing that hid orbit iteration bugs.

**Pull request merged**